pkg/block/indexheader/lazy_binary_reader.go

```diff
@@ -0,0 +1,228 @@
+		dir:                         dir,
+		filepath:                    filepath,
+		id:                          id,
+		postingOffsetsInMemSampling: postingOffsetsInMemSampling,
+		metrics:                     metrics,
+	}, nil
+}
+
+func (r *LazyBinaryReader) Close() error {
+	r.readerMx.Lock()
+	defer r.readerMx.Unlock()
+
+	if r.reader == nil {
+		return nil
+	}
+
+	r.metrics.unloadCount.Inc()
+	if err := r.reader.Close(); err != nil {
+		r.metrics.unloadFailedCount.Inc()
+		return err
+	}
+
+	r.reader = nil
+
+	return nil
+}
+
+func (r *LazyBinaryReader) IndexVersion() (int, error) {
+	r.readerMx.RLock()
+	defer r.readerMx.RUnlock()
+
+	if err := r.open(); err != nil {
+		return 0, err
+	}
+
+	return r.reader.IndexVersion()
+}
+
+func (r *LazyBinaryReader) PostingsOffset(name string, value string) (index.Range, error) {
+	r.readerMx.RLock()
+	defer r.readerMx.RUnlock()
+
+	if err := r.open(); err != nil {
+		return index.Range{}, err
+	}
+
+	return r.reader.PostingsOffset(name, value)
+}
+
+func (r *LazyBinaryReader) LookupSymbol(o uint32) (string, error) {
+	r.readerMx.RLock()
+	defer r.readerMx.RUnlock()
+
+	if err := r.open(); err != nil {
+		return "", err
+	}
+
+	return r.reader.LookupSymbol(o)
+}
+
+func (r *LazyBinaryReader) LabelValues(name string) ([]string, error) {
+	r.readerMx.RLock()
+	defer r.readerMx.RUnlock()
+
+	if err := r.open(); err != nil {
+		return nil, err
+	}
+
+	return r.reader.LabelValues(name)
+}
+
+func (r *LazyBinaryReader) LabelNames() ([]string, error) {
+	r.readerMx.RLock()
+	defer r.readerMx.RUnlock()
+
+	if err := r.open(); err != nil {
+		return nil, err
+	}
+
+	return r.reader.LabelNames()
+}
+
+// open ensure the underlying binary index-header reader has been successfully opened. Returns
+// an error on failure. This function MUST be called with the read lock already acquired.
+func (r *LazyBinaryReader) open() error {
+	// Nothing to do if we already tried opening it.
+	if r.reader != nil {
+		return nil
+	} else if r.readerErr != nil {
+		return r.readerErr
+	}
+
+	// Take the write lock to ensure we'll try to open it only once. Take again
+	// the read lock once done.
+	r.readerMx.RUnlock()
+	r.readerMx.Lock()
+	defer r.readerMx.RLock()
+	defer r.readerMx.Unlock()
+
+	// Ensure none else tried to open it in the meanwhile.
+	if r.reader != nil {
+		return nil
+	} else if r.readerErr != nil {
+		return r.readerErr
+	}
+
+	level.Debug(r.logger).Log("msg", "lazy loading index-header file", "path", r.filepath)
+	r.metrics.loadCount.Inc()
+	startTime := time.Now()
+
+	reader, err := NewBinaryReader(r.ctx, r.logger, r.bkt, r.dir, r.id, r.postingOffsetsInMemSampling)
+	if err != nil {
+		level.Error(r.logger).Log("msg", "failed to lazy load index-header file", "path", r.filepath, "err", err)
+		r.metrics.loadFailedCount.Inc()
+		r.readerErr = err
+		return err
+	}
+
+	r.reader = reader
+	level.Info(r.logger).Log("msg", "lazy loaded index-header file", "path", r.filepath, "elapsed", time.Since(startTime))
```

Comment:
Do we need this log line :thinking: Maybe it should be debug instead?

Reply:
we have nice metrics for it

Reply:
Moved to debug.